consul/src/main/java/com/linecorp/armeria/internal/consul/CatalogClient.java

```diff
@@ -0,0 +1,200 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.internal.consul;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.type.CollectionType;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.internal.common.PercentEncoder;
+
+/**
+ * A Consul client that is responsible for
+ * <a href="https://www.consul.io/api/catalog.html">Catalog HTTP API</a>.
+ */
+final class CatalogClient {
+
+    private static final CollectionType collectionTypeForNode =
+            TypeFactory.defaultInstance().constructCollectionType(List.class, Node.class);
+
+    static CatalogClient of(ConsulClient consulClient) {
+        return new CatalogClient(consulClient);
+    }
+
+    private final WebClient client;
+    private final ObjectMapper mapper;
+
+    private CatalogClient(ConsulClient client) {
+        this.client = client.consulWebClient();
+        mapper = client.getObjectMapper();
+    }
+
+    /**
+     * Gets endpoint list with service name.
+     */
+    CompletableFuture<List<Endpoint>> endpoints(String serviceName) {
+        requireNonNull(serviceName, "serviceName");
+        return service(serviceName)
+                .thenApply(nodes -> nodes.stream()
+                                         .map(CatalogClient::convertToEndpoint)
+                                         .collect(toImmutableList()));
+    }
+
+    /**
+     * Returns node list with service name.
+     */
+    @VisibleForTesting
+    CompletableFuture<List<Node>> service(String serviceName) {
+        requireNonNull(serviceName, "serviceName");
+        final StringBuilder path = new StringBuilder("/catalog/service/");
+        PercentEncoder.encodeComponent(path, serviceName);
+        return client.get(path.toString())
+                     .aggregate()
+                     .thenApply(response -> {
+                         try {
+                             return mapper.readValue(response.content().toStringUtf8(), collectionTypeForNode);
```

Comment:
I think we can use `byte[]`
`return mapper.readValue(response.content().array(), collectionTypeForNode);`

Reply:
Oh, it's good!